_posts/2021-01-05-git-rebase-merge.md

```diff
@@ -0,0 +1,202 @@
+
+It is important to note that while using the Git Merge, the current branch will be updated to reflect the merge, but the target branch remains untouched and un-reflected. The git merge is often used in combination with git checkout for the selection of the current branch, and git branch -d for deleting the obsolete target branch.
+
+## Working of Git Merge
+
+Git Merge is basically used for combining the multiple sequences of commits into one unified history. In the most frequently occurring cases, git merge is used to combine two branches. Let's take an example in which we will mainly focus on branch merging patterns. In the scenario which we have taken, the git merge takes two commit pointers, usually known as the branch tips, and we will manage to find the common base commit between them. Once the git has a common base commit, it will create a new "merge commit", that will combine the changes of each queued merge commit sequence.
+
+![Git Merge combining branches](/assets/img/posts/git-rebase-merge/git-merge-working1.png)
+
+![Git Merge with Common Base](/assets/img/posts/git-rebase-merge/git-merge-working2.png)
+
+## Example Usecase
+
+Now it is time to get a hands-on practical example. The code we are going to write will create a new branch, add two commits to it, and finally, it will integrate it to the mainline by using fast-forward merge.
+
+### Start working on a new feature
+
+```
+ git checkout -b new-feature master
+
+ # Edit some files
+
+ git add filename
+
+ git commit -m "Start a feature"
+
+ # Edit some files
+
+ git add filename
+
+ git commit -m "Finish a feature"
+
+ # Merge in the new-feature branch
+
+ git checkout master
+
+ git merge new-feature
+
+ git branch -d new-feature
+```
+
+It is to be noted that this is a common workflow for short-lived topic branches, that can be utilized for independent development as compared to an organizational tool for longer-running features.
+
+Since the new features are now accessible from the master branch, therefore, Git shouldn't complain about the git branch -d.
+
+If we require a merge commit during the fast forward merge for record-keeping purposes, then the git merge can be executed with the --no-ffoption.
+
+The command given below can be very useful if one wants to merge the specified branch into the current branch, but it is worth mentioning that it will always generate a merge commit. It can be very helpful for the documentation of all merges that occur in the repository.
+
+```
+git merge --no-ff <branch>
+```
+
+## Git Rebase vs Git Merge
+
+Now let us go through the difference between the Git Rebase and Git Merge. When we create the Git Merge the head branch will always preserve the ancestry of each commit history by generating a new commit.
+
+![Git Merge Preserving commit history](/assets/img/posts/git-rebase-merge/git-merge-history.png)

```

Comment:
If you look at both of the diagrams closely then you can see that for rebase we are not being able to preserve the commit history however for the merge we are able to preserve the history.

For the missing commit, Since the merge without any flag is a fast forward (default merge) therefore it only updates the branch pointer to match the merged branch and not create a merge commit.

I hope this makes sense now